### ISSUE TITLE
gpcheckcat owner test misses certain AO aux table/index

### ISSUE THREAD

**Issue by huansong** [greenplum-db/gpdb#15104]
## Bug Report
`gpcheckcat -R owner` checks if there is any table that has different owner on coordinator vs primaries. But it would miss the aoseg, blkdir relations and their index. For example if we do:

```sql
-- on coordinator
postgres=# create table t(a int) with(appendonly);
CREATE TABLE
postgres=# create index ti on t(a);
CREATE INDEX
postgres=# create user wolf login password 'foo';
CREATE ROLE

-- on a primary, in utility mode
postgres=# alter table t owner to wolf;
ALTER TABLE
-- check the OIDs of the AO table and all of its aux table/index:
postgres=# select ao.* from pg_appendonly ao, pg_class c where ao.relid = c.oid and c.relname = 't';
 relid | segrelid | blkdirrelid | visimaprelid
-------+----------+-------------+--------------
 54805 |    54807 |       54812 |        54809
postgres=# select indexrelid from pg_index where indrelid in (54805, 54807, 54812, 54809);
 indexrelid
------------
      54811
      54814
      54815
(3 rows)
-- can see that the owner of the AO table and all of its aux table/index get changed:
postgres=# select c.relname, a.rolname from pg_class c left join pg_authid a on (a.oid = c.relowner) 
where c.oid in (54805, 54807, 54812, 54809, 54811, 54814, 54815);
         relname          | rolname
--------------------------+---------
 t                        | wolf
 pg_aoseg_54805           | wolf
 pg_aovisimap_54805       | wolf
 pg_aovisimap_54805_index | wolf
 pg_aoblkdir_54805        | wolf
 pg_aoblkdir_54805_index  | wolf
 ti                       | wolf
(7 rows)
```

Now if we do `gpcheckcat -R owner postgres`, it can only identify the AO table, index and the aovisimap table/index, but miss the rest. See the repair script it generates:
```sql
-- owner for "pg_aoseg"."pg_aovisimap_54805"
ALTER TABLE "pg_aoseg"."pg_aovisimap_54805" OWNER TO "wolf";
ALTER TABLE "pg_aoseg"."pg_aovisimap_54805" OWNER TO "csteam";
-- owner for "pg_aoseg"."pg_aovisimap_54805_index"
ALTER TABLE "pg_aoseg"."pg_aovisimap_54805_index" OWNER TO "wolf";
ALTER TABLE "pg_aoseg"."pg_aovisimap_54805_index" OWNER TO "csteam";
-- owner for "public"."t"
ALTER TABLE "public"."t" OWNER TO "wolf";
ALTER TABLE "public"."t" OWNER TO "csteam";
-- owner for "public"."ti"
ALTER TABLE "public"."ti" OWNER TO "wolf";
ALTER TABLE "public"."ti" OWNER TO "csteam";
```

Found this in https://github.com/greenplum-db/gpdb/pull/15100 when fixing some issue with the query, but this bug is unrelated to that change. The problem is with the query in gpcheckcat itself (https://github.com/greenplum-db/gpdb/blob/main/gpMgmt/bin/gpcheckcat#L1068-L1081), which might be outdated (or wrong in the first place?) Just need to rewrite the query properly, and maybe add a behave test that checks the same.

### Greenplum version or build
main branch (HEAD 0f243b2ba5c548c57408980d026d6a3767c83518)
6X is the same.

**Comment by hyongtao-db** [comment]
@huansong  Hi, could you please help to review [this pr](https://github.com/greenplum-db/gpdb/pull/15453)  ? No hurry, just a reminder. Thanks a lot.

**Comment by hyongtao-db** [comment]
Seems it is so hard to add test case, so I record the results of manual tests here:

**Comment by huansong** [comment]
@hyongtao-db Thanks for working on this! 

As I took another look at the problematic query, I came to realize that it seems intentional that gpcheckcat doesn't show or fix AO aux tables: the AO aux tables should be fixed with their main AO table. For example when we do 

```sql
alter table t owner to wolf;
```

All of its aux tables (aoseg, aoblkdir, aoblkdiridx, aovisimap, aovisimapidx) should all be changed to have the same owner. In fact, we are not allowed to alter the owner of the aux table themselves:
```sql
postgres=# alter table pg_aoseg.pg_aoseg_32795 owner to wolf;
ERROR:  permission denied: "pg_aoseg_32795" is a system catalog
```

Basically, AO aux tables should be handled in the same way as toast table. And as we can see from that [problematic query](https://github.com/greenplum-db/gpdb/pull/15453/files#diff-ecfa56ea883f96bae7a65630548f02ad4a6c2b43da05966cb6ca10fd9d85cc8cL1077), toast tables are also ignored in the result.

Therefore, now I think the real issue is not that "_aoseg/aoblkdir/aoblkdiridx are not fixed_", but instead "_aovisimap/aovisimapidx are fixed_".  We should ignore aovisimap/aovisimapidx in the same way as the other aux tables. 

So if you agree, let's fix the query as (I haven't tested it):
```sql
select distinct n.nspname, coalesce(o.relname, c.relname) as relname,
                    a.rolname, m.rolname as coordinator_rolname
    from gp_dist_random('pg_class') r
      join pg_class c on (c.oid = r.oid)
      left join pg_index i on (c.oid = i.indexrelid)
      left join pg_appendonly ao on (c.oid = ao.segrelid or
                                     c.oid = ao.blkdirrelid or
                                     c.oid = ao.visimaprelid or
                                     i.indrelid = ao.blkdirrelid or
                                     i.indrelid = ao.visimaprelid)
      left join pg_class o on (o.oid = ao.relid or
                               o.reltoastrelid = c.oid)
      join pg_authid a on (a.oid = r.relowner)
      join pg_authid m on (m.oid = coalesce(o.relowner, c.relowner))
      join pg_namespace n on (n.oid = coalesce(o.relnamespace, c.relnamespace))
    where c.relowner <> r.relowner
```

**Comment by hyongtao-db** [comment]
> Therefore, now I think the real issue is not that "aoseg/aoblkdir/aoblkdiridx are not included", but instead "aovisimap/aovisimapidx are included". We should ignore aovisimap/aovisimapidx in the same way as the other aux tables.

Agree with you. Thanks for your deatailed explaination. Please let me test it based on the new query.

**Comment by hyongtao-db** [comment]
> select distinct n.nspname, coalesce(o.relname, c.relname) as relname,
                    a.rolname, m.rolname as coordinator_rolname
    from gp_dist_random('pg_class') r
      join pg_class c on (c.oid = r.oid)
      left join pg_index i on (c.oid = i.indexrelid)
      left join pg_appendonly ao on (c.oid = ao.segrelid or
                                     c.oid = ao.blkdirrelid or
                                     c.oid = ao.visimaprelid or
                                     i.indrelid = ao.blkdirrelid or
                                     i.indrelid = ao.visimaprelid)
      left join pg_class o on (o.oid = ao.relid or
                               o.reltoastrelid = c.oid)
      join pg_authid a on (a.oid = r.relowner)
      join pg_authid m on (m.oid = coalesce(o.relowner, c.relowner))
      join pg_namespace n on (n.oid = coalesce(o.relnamespace, c.relnamespace))
    where c.relowner <> r.relowner

The test results base on this query is shown as followed:

**Comment by hyongtao-db** [comment]
This issue is fixed in #15453 . 
Thanks the great help from @huansong and @higuoxing .
Close this issue.
